constraint/pkg/externaldata/cache.go

```diff
@@ -16,6 +18,74 @@ type ProviderCache struct {
 	mux   sync.RWMutex
 }
 
+type ProviderResponseCache struct {
+	cache        map[CacheKey]CacheValue
+	mux          sync.RWMutex
+	TTLInMinutes time.Duration
+}
+
+type CacheKey struct {
+	ProviderName string
+	Key          string
+}
+
+type CacheValue struct {
+	Received int64
+	Value    interface{}
+	Error    string
+}
+
+func NewProviderResponseCache(ttlInMinutes time.Duration, ctx context.Context) *ProviderResponseCache {
+	providerResponseCache := &ProviderResponseCache{
+		cache:        make(map[CacheKey]CacheValue),
+		TTLInMinutes: ttlInMinutes,
+	}
+
+	go providerResponseCache.invalidateProviderResponseCache(ctx, ttlInMinutes)
+
+	return providerResponseCache
+}
+
+func (c *ProviderResponseCache) Get(key CacheKey) (*CacheValue, error) {
+	c.mux.RLock()
+	defer c.mux.RUnlock()
+
+	if v, ok := c.cache[key]; ok {
+		return &v, nil
+	}
+	return nil, fmt.Errorf("key '%s:%s' is not found in provider response cache", key.ProviderName, key.Key)
+}
+
+func (c *ProviderResponseCache) Upsert(key CacheKey, value CacheValue) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	c.cache[key] = value
+}
+
+func (c *ProviderResponseCache) Remove(key CacheKey) {
+	c.mux.Lock()
+	defer c.mux.Unlock()
+
+	delete(c.cache, key)
+}
+
+func (c *ProviderResponseCache) invalidateProviderResponseCache(ctx context.Context, ttl time.Duration) {
+	for {
```

Comment:
Is there any throttling of this for-loop (a sleep, a timer tic, etc.)?

Reply:
Updated code